[WIP] Initial, buildable pass

eriknelson opened the pull request:
Features:
* Removed `Client` and `NewClient` from apb/client.go, wasn't doing much for us, especially with the new clients package. Required much refactoring.
* Renamed RunImage to ExecuteApb. With `Client` gone, turned into function instead of method. Only reason it was a method before was to get access to the log set on the struct, not useful. apb/client.go becomes apb/executor.go. Happy to see it's become a **lot** smaller, cleaner, and more direct about what it's doing.
* `clients` package refactorings.
  * Made instances struct private, instances should be retrieved via `client.<instance>(...)`. Cleaner than `clients.Clients.<instance>`.
  * Singleton thread safety via `sync.Once`
  * All client ctors are private and internal to clients package.
* Reverted to use DaoConfig with embedded EtcdConfig
* Removed PullImage. We can likely go farther than this by shedding docker dep **TODO: link issue**
* Moved OcLogin out of apb package into broker package and made it private, it was the only place using it.
* GetEtcdVersion moved to function instead of method. This will probably change anyway with the officially supported ETCD version endpoint change **TODO: link PR**.
* Removed redundant Client creations from app.Start that don't really do anything becauses force instantiation. If we want to do this, we should have an explicitly supported way in clients package, I.E. `clients.Init()`.
* Pulled RESTClient out of Clients. It's a derivative of k8s clientset, I don't think it needs to be on the top level struct.

TODO:
* Singletons in each of the clients are algorithmically doing the **exact** same thing, just varying by type and possibly client ctor signatures. Need to figure out a way to DRY this up.
* ~~<instance>ClientResult objects are also literally the same, varying by client type. It's just a client/error pair. Need something like `Result<ClientType>()`.~~ Moved to `clientResult`, still requires type assertions, not super happy with this either.

Commit:
Break apart the Kubernetes Client

Make a seperate directory for clients so they are
not apart of the APB code.  This will make the clients
easy to initialize and create new ones.

Commit:
Break apart the Docker Client

Commit:
Break apart the Etcd client

The Etcd Client is a little tricker because it requires
a config file. Later when we unify all the clients we will
use individual functions just like etcd.

Commit:
Create a global client variable

Pull together all the client split work by using a global variable
to pass around the clients.

eriknelson committed:
Initial, buildable pass

eriknelson committed:
Rename apb/client.go to apb/executor.go

eriknelson committed:
Remove stale comments

eriknelson committed:
Remove type specific client results

eriknelson commented:
Think some things are busted with clientResults

eriknelson committed:
Fix build errors with type assertion check

eriknelson committed:
Remove InCluster branch from executor

rthallisey commented on pkg/apb/executor.go:
```diff
@@ -0,0 +1,110 @@
+	Password  string `yaml:"pass"`
+}
+
+func ExecuteApb(
```
These are public functions.  Maybe they should be part of the apb class or in their own class/pkg?

jmrodri replied:
I agree with @rthallisey here. Having this in executor feels weird. Though not sure how I would do it :)

eriknelson replied:
Mentioned this to zeus, I'm picturing this function effectively turning into something much more generic that delegates the actual running of the apb to a cluster specific runner that implements an interface.

rthallisey commented on pkg/apb/provision.go:
```diff
@@ -55,17 +55,13 @@ func Provision(
 
 	if err != nil {
-		log.Error("Problem running image")
+		log.Error("Problem executing apb")
```
This error is "Problem executing apb" then the next line is the name. How about adding podName to "Problem executing apb: %s".

You could also keep this as running.

jmrodri replied:
Should probably do that for all of them.

rthallisey commented on pkg/clients/clients.go:
```diff
@@ -1,16 +1,22 @@
+var once struct {
+	Etcd       sync.Once
+	Kubernetes sync.Once
+	Docker     sync.Once
```
👍

rthallisey reviewed (commented):
Neat addition @eriknelson. I'll review it more tomorrow, but I like the client distribution.

jmrodri committed:
techdebt: speed up builds (#234)

jmrodri commented on pkg/dao/dao.go:
```diff
@@ -41,10 +49,11 @@ func (d *Dao) SetRaw(key string, val string) error {
 	// We'll use this to test our etcd connection for now
-	resp, err := http.Get("http://" + config.EtcdHost + ":" + config.EtcdPort + "/version")
+	etcdUrl := fmt.Sprintf("http://%s:%s/version", ec.EtcdHost, ec.EtcdPort)
+	resp, err := http.Get(etcdUrl)
```
Except for the long line, I would've just put them in one line.

```
resp, err := http.Get(fmt.Sprintf("http://%s:%s/version", ec.EtcdHost, ec.EtcdPort))
```

eriknelson replied:
Personal style habit. I don't like too must cruft on one line :)

jmrodri commented on pkg/apb/bind.go:
```diff
@@ -25,17 +25,13 @@ func Bind(
+	podName, err := ExecuteApb(
+		"bind", clusterConfig, instance.Spec,
+		instance.Context, parameters, log,
+	)
```
This makes sense. I do like the `client.` though since it was explicit we were doing something on the client. But meh this is ok.

jmrodri replied:
Would be neat if it were `ExecuteBind(clusterConfig, ....)` but that's a nit and probably not worth all the indirection.

eriknelson replied:
+1, I was musing over this earlier. Should probably define a finite amount of typed actions ExecuteApb actually accepts so you can't build a broker that feeds garbage into ExecuteApb.

jmrodri commented on pkg/apb/bind.go:
```diff
@@ -25,17 +25,13 @@ func Bind(
 
 	if err != nil {
-		log.Error("Problem running image", err)
+		log.Error("Problem executing apb: %s", err)
```
good log cleanup

jmrodri commented on pkg/apb/client.go:
```diff
@@ -1,188 +0,0 @@
-	log           *logging.Logger
-}
-
-func NewClient(log *logging.Logger) (*Client, error) {
```
:+1: :clap:

eriknelson commented on pkg/apb/deprovision.go:
```diff
@@ -27,19 +27,14 @@ func Deprovision(instance *ServiceInstance, clusterConfig ClusterConfig, log *lo
-	podName, err := client.RunImage(
-		"deprovision", clusterConfig, instance.Spec, instance.Context, instance.Parameters)
+	podName, err := ExecuteApb(
+		"deprovision", clusterConfig, instance.Spec,
```
Looking this over again, probably should make that first arg a typed "ApbAction" enum.

jmrodri commented on pkg/apb/executor.go:
```diff
@@ -0,0 +1,110 @@
+		return "", errors.New(errStr)
+	}
+
+	ns := context.Namespace
```
why copy it?

eriknelson replied:
Simple alias.

jmrodri commented on pkg/apb/executor.go:
```diff
@@ -0,0 +1,110 @@
+	}
+
+	log.Notice(fmt.Sprintf("Creating pod %q in the %s namespace", pod.Name, ns))
+	k8scli, err := clients.Kubernetes(log)
```
How would this function use a different client?

jmrodri replied:
Can we use an interface for the clients? I've seen many have a `Do()` method. 
```
type ClientInterface interface {
    Do() (interface{}, error)
}
```

eriknelson replied:
@jmrodri I think our heads our in the same place. I'm envisioning most of the logic inside `ExecuteApb` to end up implemented behind an interface that cluster specific types implement. Factory method allocs/inits the runner and `ExecuteApb` effectively just calls runner.Run(args).

Influenced by this parallel discussion: https://github.com/fusor/apb-examples/issues/60#issuecomment-311073195

jmrodri commented on pkg/app/app.go:
```diff
@@ -63,34 +63,35 @@ func CreateApp() App {
+	// Something like, clients.ValidateEtcd()
 	app.log.Debug("Connecting to Etcd")
-	serv, clust, err := app.dao.GetEtcdVersion(app.config.Dao)
+	serv, clust, err := dao.GetEtcdVersion(app.config.Dao.EtcdConfig)
```
+1 to using the `config.Dao`

jmrodri commented on pkg/broker/broker.go:
```diff
@@ -769,3 +769,17 @@ func (a AnsibleBroker) RemoveSpecs() error {
+		return err
+	}
+	return nil
+}
```
What would this look like if it were an origin client?

eriknelson replied:
Not sure yet, setting up an origin client hello world is on my TODO list, I've been unsuccessful at finding an example.

More cluster specifics we'll need to abstract away too. `client.Login` interface method?

jmrodri commented on pkg/clients/docker.go:
```diff
@@ -1,17 +1,37 @@
+			log.Error(err.Error())
+			instances.Docker = clientResult{nil, err}
+		}
+		instances.Docker = clientResult{client, nil}
```
This is actually pretty nice. I can handle this. :+1:

eriknelson replied:
I tried to reorg these into something generic. It eventually just became the once method with a bunch of type conversions. Almost no gain there, it's error handling boilerplate for the most part that would be required anyway, so I abandoned it. :/

jmrodri reviewed (approved):
Tests would be AWESOME! Overall I like it.

jmrodri committed:
remove mockregistry (#236)

* remove the mock-registry service

We used to have a mock-registry service you could run with the broker
that was driven by a data file. This service ran seperately from the
broker.

This is NOT the same as the current mock_registry located IN the broker.

* remove -registry left over

* remove dev_registry which connected to mock_reg

Commit:
Breakup all the Broker Clients into a clients pkg (#222)

* Break apart the Kubernetes Client

Make a seperate directory for clients so they are
not apart of the APB code.  This will make the clients
easy to initialize and create new ones.

* Break apart the Docker Client

* Break apart the Etcd client

The Etcd Client is a little tricker because it requires
a config file. Later when we unify all the clients we will
use individual functions just like etcd.

* Create a global client variable

Pull together all the client split work by using a global variable
to pass around the clients.
